pkg/placement/placement-decision.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2023 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package placement
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/kubestellar/kubestellar/api/edge/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+// syncPlacementDecision syncs a placement-decision object with what is resolved by the placement-resolver.
+func (c *Controller) syncPlacementDecision(key util.Key) error {
+	var unstructuredObj *unstructured.Unstructured
+
+	if obj, err := c.getObjectFromKey(key); err != nil {
+		unstructuredObj = util.EmptyUnstructuredObjectFromKey(key)
+	} else {
+		// perform the type assertion only if getObjectFromKey did not fail
+		var ok bool
+		unstructuredObj, ok = obj.(*unstructured.Unstructured)
+		if !ok {
+			return fmt.Errorf("failed to convert runtime.Object to unstructured.Unstructured")
+		}
+	}
+
+	placementDecision, err := unstructuredObjectToPlacementDecision(unstructuredObj)
+	if err != nil {
+		return fmt.Errorf("failed to convert from Unstructured to PlacementDecision: %w", err)
+	}
+
+	// placement decision name matches that of the placement 1:1, therefore its NamespacedName is the same.
+	placementIdentifier := placementDecision.GetName()
+
+	// generate placement decision spec from resolver
+	generatedPlacementDecisionSpec, err := c.placementResolver.GeneratePlacementDecision(placementIdentifier)
+	if err != nil {
+		return fmt.Errorf("failed to get placement decision spec: %w", err)
+	}
+
+	// calculate if the resolved decision is different from the current one
+	if !c.placementResolver.ComparePlacementDecision(placementIdentifier, &placementDecision.Spec) {
+		// update the placement decision object in the cluster by updating spec
+		if err = c.updateOrCreatePlacementDecision(placementDecision, generatedPlacementDecisionSpec); err != nil {
```

Comment:
FYI. A false negative from `ComparePlacementDecision` due to lack of gvk/gvr mapping will likely cause an error to be returned here for the same reason. I think it would be better to make `ComparePlacementDecision` return an error in that case and save the doomed work here.

Reply:
I think I'll keep it as it currently is, as this also falls under something that would be resolved by reworking the Key to include object Resource. Is that ok with you?

Reply:
yes.